documentation/optee_design.md

```diff
@@ -487,6 +487,88 @@ implementation, as explained in [crypto.md].
 TBD
 
 # 11. Trusted Applications
+
+## Static and Dynamic TAs
+
+There are two ways to implement Trusted Applications, static and dynamic.
+For most cases, Dynamically-loaded TAs are preferred.
+
+### Static Trusted Applications
+
+These are added directly to the optee_os tree in, eg, `core/arch/arm/sta`, and
+are built along with and built into the OS blob.
+
+You can look at `core/arch/arm/sta/stats.c` as a template and just add your
+static TA based on that to the `sub.mk` in the same directory.
+
+Static TAs have the same privileged view of Secure World as the optee_os core
+itself.  For complex TAs, that may not be desirable.
+
+### Dynamic Trusted Applications
+
+Trusted Applications consists of a signed ELF file.
+
+They are built separately from the optee_os boot-time blob, although when they
+are built they use the same build system, and are signed with the key from the
+build of the original optee_os blob.
+
+OP-TEE is able to dynamically load these TA files, along the same lines as the
+Linux kernel loading a module.  And just as the Linux kernel can confirm at runtime
+that the module was signed by the same build as the main kernel binary, OP-TEE
+checks the TA file signature using the public key from when the main optee_os blob
+was built itself.
+
+This allows the signed TAs to be stored on untrusted storage in normal world, and
+loaded on demand when a client requests its UUID.
+
+## Special treatment of Dynamic TAs
+
+### Syscalls
+
+Since Dynamic TAs are not directly bound to function exports in the optee_os blob,
+since that would require upgrades of all TAs synchronusly to upgrades of the
+optee_os blob, the use of optee_os exports by the Dynamic TA must be resolved
+at runtime.
+
+optee_os does this by using syscalls, the same kind of way as the Linux kernel
+provides a stable API for its userland programs.  TAs are written to use
+syscall wrappers to access functions exported from optee_os, so this all
+happens automatically when a TA wants to use an api exported from core optee_os.
+
+Static TAs and anything else in the core optee_os blob do not require going
+through a syscall interface, since they can just link directly as they are
+part of the core blob.
+
+Syscalls are provided already for all public exports from the core that a
+Dynamic TA is expected to use, so you only need to take care about this if
+you will add new exported from the core that TAs will want to use.
+
+### Malloc mapping
+
+When the core optee_os performs a `malloc()`, the VA mapping is placed in the
+MMC table only for core optee_os threads.  This means that TAs cannot observe
+data that is being allocated and used by the core optee_os.
```

Comment:
i would rather remove this as it's not TA related.
Rather clearly state that the preferred method for memeory allocation is, as specified by the GP specs, the TEE_Malloc() and friends.